src/libfsm/collect.c

```diff
@@ -0,0 +1,108 @@
+#include <assert.h>
+#include <stdlib.h>
+#include <errno.h>
+
+#include <adt/queue.h>
+#include <adt/set.h>
+#include <adt/edgeset.h>
+#include <adt/stateset.h>
+
+#include "internal.h"
+
+static void
+clear_states(struct fsm_state *s)
+{
+	for (; s != NULL; s = s->next) { s->reachable = 0; }
+}
+
+static int
+mark_states(struct fsm *fsm)
+{
+	const unsigned state_count = fsm_countstates(fsm);
+	struct queue *q = queue_new(state_count);
+	int res = 0;
+	if (q == NULL) { return 1; }
+
+	if (!queue_push(q, fsm->start)) {
+		goto cleanup;
+	}
+	fsm->start->reachable = 1;
+	
+	for (;;) {
```

Comment:
Maybe worth generalizing this loop into a general BFS walk?

Reply:
Kate and I talked about doing that for a couple other functions as well, but probably in a different PR.

Reply:
That makes sense.

Incidentally, I'm curious why you went with BFS here instead of a recursive DFS.  The BFS would be useful for Dijkstra's shortest path and similar algs, but for mark-and-sweep, you just want to visit reachable nodes, and a DFS should handle that.

Reply:
Whether it's BFS or DFS is pretty arbitrary to me, my main concern was that it didn't overrun the stack when marking a huge graph recursively.